doc/sphinx/proof-engine/ltac.rst

```diff
@@ -1661,6 +1661,100 @@ Testing boolean expressions: guard
    .. exn:: Condition not satisfied.
       :undocumented:
 
+Checking properties of terms
+----------------------------
+
+Each of the following tactics acts as the identity if the check
+succeeds, and results in an error otherwise.
+
+.. tacn:: constr_eq @one_term @one_term
+
+   Checks whether its arguments are equal ignoring differences in alpha
+   conversion, casts and universe constraints. It may unify universes.
+
+   .. exn:: Not equal.
+      :undocumented:
+
+   .. exn:: Not equal (due to universes).
+      :undocumented:
+
+.. tacn:: constr_eq_strict @one_term @one_term
+
+   This tactic checks whether its arguments are equal modulo alpha
+   conversion, casts and universe constraints. It does not add new
+   constraints.
+
+   .. exn:: Not equal.
+      :undocumented:
+
+   .. exn:: Not equal (due to universes).
+      :undocumented:
+
+.. tacn:: constr_eq_nounivs @one_term @one_term
+
+   Checks whether its arguments are equal modulo alpha
+   conversion, casts and universe constraints. It does not add new
+   constraints.
+
+.. tacn:: unify @one_term @one_term {? with @ident }
+
+   Checks whether its arguments are unifiable, potentially
+   instantiating existential variables.
+
+   Unification takes the transparency information defined in the hint database
+   :n:`@ident` into account (see :ref:`the hints databases for auto and eauto
+   <thehintsdatabasesforautoandeauto>`).
+
+.. tacn:: is_evar @term
+   :name: is_evar
+
+   Checks whether its argument is a current existential
+   variable. Existential variables are uninstantiated variables generated
+   by :tacn:`eapply` and some other tactics.
+
+   .. exn:: Not an evar.
+      :undocumented:
+
+.. tacn:: has_evar @term
+   :name: has_evar
+
+   Checks whether its argument has an existential variable as
+   a subterm. Unlike context patterns combined with ``is_evar``, this tactic
+   scans all subterms, including those under binders.
+
+   .. exn:: No evars.
+      :undocumented:
+
+.. tacn:: is_var @term
+   :name: is_var
+
+   Checks whether its argument is a variable or hypothesis in
+   the current local context.
+
+   .. exn:: Not a variable or hypothesis.
+      :undocumented:
+
+.. tacn:: is_cofix @one_term
+   :undocumented:
+
+.. tacn:: is_const @one_term
+   :undocumented:
+
+.. tacn:: is_constructor @one_term
+   :undocumented:
+
+.. tacn:: is_fix @one_term
+   :undocumented:
+
+.. tacn:: is_ground @one_term
+   :undocumented:
```

Comment:
As far as I can tell, this is equivalent to `assert_fails (has_evar @one_term)`, but I'm not completely sure.

Reply:
And the error message is "Not ground.", and can be obtained by, e.g., `is_ground open_constr:(?[c]).`, while something like `is_ground 0` succeeds.

Reply:
This is the underlying code.  Is there a user-visible difference between defined and undefined evars?  The "undefined" distinction may just be internals.  Perhaps you could construct an example to test whether there's a difference?  If they are different, I'd like to explain that very briefly.

```
let has_undefined_evars evd t =
  let rec has_ev t =
    match EConstr.kind evd t with
    | Evar _ -> raise NotInstantiatedEvar
    | _ -> EConstr.iter evd has_ev t in
  try let _ = has_ev t in false
  with (Not_found | NotInstantiatedEvar) -> true

let is_ground_term evd t =
  not (has_undefined_evars evd t) 
```

Reply:
`is_ground` is in fact implemented as the negation of `has_evar`.  The "undefined" part is just in naming, and is implicit in the fact that we're using EConstr, where all observable evars are undefined.  (In Constr-land we can have evars that happen to be defined already.)
https://github.com/coq/coq/blob/cf60242ed3022c332d3de88fbcbab4587cda83b5/plugins/ltac/internals.ml#L299-L303
https://github.com/coq/coq/blob/cf60242ed3022c332d3de88fbcbab4587cda83b5/tactics/class_tactics.ml#L1389-L1393
https://github.com/coq/coq/blob/cf60242ed3022c332d3de88fbcbab4587cda83b5/engine/evarutil.ml#L99-L108